### PR TITLE
Do not override existing web.config during IIS publish

### PR DESCRIPTION
**Pull request by captainsafia**
Only copies standard web.config if one does not already exist in the publish directory.

Addresses #20490

### CODE REVIEW

**Review by javiercn — APPROVED** [review]
Looks great!

Nothing to add here. The fix is exactly what I expected.

Great job!

File: src/Components/WebAssembly/Build/test/BuildIntegrationTests/PublishIntegrationTest.cs [hunk 1]
```diff
@@ -101,6 +101,38 @@ public async Task Publish_InRelease_Works()
...
+        {
+            // Arrange
+            using var project = ProjectDirectory.Create("standalone", additionalProjects: new[] { "razorclasslibrary", "LinkBaseToWebRoot" });
+            project.Configuration = "Release";
+
+            var webConfigContents = @"<?xml version='1.0' encoding='utf-8' ?>
+< configuration >
+  < system.web >
+  </ system.web >
+</ configuration >";
```
Comment (SteveSandersonMS) [inline src/Components/WebAssembly/Build/test/BuildIntegrationTests/PublishIntegrationTest.cs:115]: Looks like VS reformatted this for you :)

I would suggest either:

 * If this is meant to be a valid `web.config` file, remove the extra spaces
 * Or if it doesn't have to be valid, replace the whole thing with something like `var webConfigContents = "test web.config contents";`

File: src/Components/WebAssembly/Build/test/BuildIntegrationTests/PublishIntegrationTest.cs [hunk 2]
```diff
@@ -101,6 +101,38 @@ public async Task Publish_InRelease_Works()
...
+            AddFileToProject(project, "web.config", webConfigContents);
+
+            AddSiblingProjectFileContent(project, @"
+<Target Name='CopyWebConfigOnPublish' AfterTargets='Publish'>
+  <Copy SourceFiles='web.config' DestinationFolder='$(PublishDir)' />
+</Target> ");
```
Comment (SteveSandersonMS) [inline src/Components/WebAssembly/Build/test/BuildIntegrationTests/PublishIntegrationTest.cs:121]: I don't quite get what this is for. Don't our publish targets already take care of copying the web.config file? Why did you need to add this target?

File: src/Components/WebAssembly/Build/src/targets/Publish.targets [hunk 3]
```diff
@@ -24,7 +24,8 @@
 
   <Target
       Name="_BlazorCopyStandaloneWebConfig"
-      AfterTargets="_BlazorCleanupPublishOutput;ComputeResolvedFilesToPublishList">
+      AfterTargets="_BlazorCleanupPublishOutput;ComputeResolvedFilesToPublishList"
+      Condition="!Exists('$(PublishDir)web.config')">
```
Comment (pranavkm) [inline src/Components/WebAssembly/Build/src/targets/Publish.targets:28]: Doesn't this assume that the web.config already exists in the publish directory before publishing has copied files over?

File: src/Components/WebAssembly/Build/test/BuildIntegrationTests/PublishIntegrationTest.cs [hunk 4]
```diff
@@ -391,6 +423,12 @@ private static void AddSiblingProjectFileContent(ProjectDirectory project, strin
             var updated = existing.Replace("<!-- Test Placeholder -->", content);
             File.WriteAllText(path, updated);
         }
+        private static void AddFileToProject(ProjectDirectory project, string filename, string content)
```
Comment (pranavkm) [inline src/Components/WebAssembly/Build/test/BuildIntegrationTests/PublishIntegrationTest.cs:426]: Needs a new line prior to this. Alternatively, just inline this

File: src/Components/WebAssembly/Build/src/targets/Publish.targets (hunk 3 above)
Comment (SteveSandersonMS) [reply src/Components/WebAssembly/Build/src/targets/Publish.targets:28]: Hmm, it's a good question, but if that's the case wouldn't our tests be failing in CI? They should be publishing from a clean state.

**Comment by captainsafia** [comment]
Thanks for the feedback everyone! The original PR was off-base in a few ways. I did more digging into this and here is what I found.

> I don't quite get what this is for. Don't our publish targets already take care of copying the web.config file? Why did you need to add this target?


Yes, the publish targets do take care of copying the web.config over. However, when I was debugging this I was seeing that any web.config files were not being copied over. As it turns out, the `_BlazorCleanupPublishOutput` was removing the `web.config` assets from the list of files that would be copied over to the `publish/` directory.

To resolve this, I changed the conditions in `_BlazorCleanupPublishOutput` to be less of a catch-all, opting to specifically remove `*.pdb`, `*.dll`, and the `Excluded Static Assets` directory. This way, any `*.config` files that the user had would still be copied over.

Following this, I made some updates to the `_BlazorCopyStandaloneWebConfig`. Now, it runs after the resolved files have been copied to publish and copies the contents of the standalone web config using `WriteLinesToFile`.

**Comment by captainsafia** [comment]
Closing this in favor of #20800.
